### ISSUE TITLE
React-specific rules

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#576]
Some ideas:

- require `displayName` property on classes that extend `React.Component`
  - https://github.com/palantir/tslint-react/issues/1
- `member-ordering` option that orders methods based on lifecycle
  - https://github.com/palantir/tslint-react/issues/2
- prohibit properties / methods on React components:
  - `propTypes` (TS should manage typechecking, not React)
    - https://github.com/palantir/tslint-react/issues/3
  - `mixins` (?)
  - `getInitialState` (state should just be initialized on the public class property)
- multiline JSX should be wrapped in parens
  - https://github.com/palantir/tslint-react/issues/5
- single line JSX should _not_ be wrapped in parens
- ban multiline JS expressions (`{ ... }`) in JSX
  - https://github.com/palantir/tslint-react/issues/4
- self-closing tags should self close (`<Logo />`, not `<Logo></Logo>`)
  - https://github.com/palantir/tslint-react/issues/6
- JSX attribute alignment (https://github.com/palantir/tslint-react/issues/7)

  ```ts
  <input
      type="search"
      placeholder="Search..."
  />
  ```
- JSX attribute sorting (alphabetize)
  - https://github.com/palantir/tslint-react/issues/8
- forbid `ReactDOM.findDOMNode()` within a component class
  - actually this can be done with the `ban` rule
- forbid certain JSX attributes, such as `dangerouslySetInnerHTML`
  - https://github.com/palantir/tslint-react/issues/9
- forbid the `on` prefix for method names in react component classes -- prefer the prefix `handle` instead, and use `on` in JSX prop names only
  - https://github.com/palantir/tslint-react/issues/10
- [x] forbid string literal refs (instead use the callback variety of refs)
  - done in palantir/tslint-react

reference:
- http://reactjsnews.com/react-style-guide-patterns-i-like/
- https://facebook.github.io/react/docs/working-with-the-browser.html

**Comment by nikolaymatrosov** [comment]
+1

**Comment by thasner** [comment]
+1 to all

**Comment by noamkeidar** [comment]
+1

Would love to see similar rules to AirBnb’s: https://github.com/airbnb/javascript/tree/master/react

**Comment by JKillian** [comment]
Another rule idea (I added it to the checklist above):

`no-string-refs`: bans string literal refs 

String literal refs [are considered a legacy feature](https://facebook.github.io/react/docs/more-about-refs.html#the-ref-string-attribute) and may be deprecated soon

**Comment by adidahiya** [comment]
For anyone following this thread:

We've started a new repo called __tslint-react__: https://github.com/palantir/tslint-react

This will contain rules related to react & JSX. I've gone ahead and filed the rule suggestions above as issues in that repo -- follow along there for updates! We should have an initial release on NPM shortly.
